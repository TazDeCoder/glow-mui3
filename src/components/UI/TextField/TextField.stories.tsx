import type { ComponentMeta, ComponentStory } from '@storybook/react';

import { TextField } from './TextField';
import { Flex } from '../../Utils';

export default {
  title: 'Components/UI/TextField',
  component: TextField,
  argTypes: {
    onFocus: { action: 'focused' },
  },
} as ComponentMeta<typeof TextField>;

export const Default: ComponentStory<typeof TextField> = (args) => (
  <TextField label="Label" {...args} />
);

export const Colors: ComponentStory<typeof TextField> = (args) => (
  <Flex flexWrap="wrap" gap={2}>
    <TextField {...args} color="primary" label="Primary" />
    <TextField {...args} color="secondary" label="Secondary" />
    <TextField {...args} color="info" label="Info" />
    <TextField {...args} color="error" label="Error" />
    <TextField {...args} color="warning" label="Warning" />
    <TextField {...args} color="success" label="Success" />
  </Flex>
);

export const Variants: ComponentStory<typeof TextField> = (args) => (
  <Flex gap={2}>
    <TextField {...args} variant="filled" label="Filled" />
    <TextField {...args} variant="outlined" label="Outlined" />
  </Flex>
);
