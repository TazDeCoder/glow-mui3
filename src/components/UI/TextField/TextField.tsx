/** @jsxRuntime classic */
/** @jsx jsx */

import { forwardRef } from 'react';
import { jsx, ThemeUICSSObject, ThemeUIStyleObject } from '@theme-ui/core';

import { Box, BoxOwnProps, BoxProps } from '../../Utils';
import { Typography } from '../Typography';
import { __internalProps } from '../../shared/utils';
import type { Assign, ForwardRef, ColorTheme } from '../../types';

export interface TextFieldProps
  extends Assign<React.ComponentPropsWithRef<'input'>, BoxOwnProps> {
  /** Color theme used for the component
   * @default 'primary'
   */
  color?: ColorTheme;
  /** Variant style of the component
   * @default 'filled'
   */
  variant?: 'filled' | 'outlined';
  /** Label associated to the component */
  label?: string;
  /** Disabled state of the component
   * @default false
   */
  disabled?: boolean;
}

const __labelCss: ThemeUICSSObject = {
  position: 'relative',
  display: 'flex',
  flexWrap: 'wrap',
  alignItems: 'center',
  width: 'min-content',
  height: 56,
  px: 16,
  overflow: 'hidden',
  bg: 'surface.variant',
  borderBottom: '1px solid currentColor',
  borderTopLeftRadius: 'xs',
  borderTopRightRadius: 'xs',
  '&:hover::after': {
    content: '""',
    position: 'absolute',
    top: 0,
    left: 0,
    width: '100%',
    height: '100%',
    bg: 'action.hover',
    borderRadius: 'inherit',
  },
};

export const TextField: ForwardRef<HTMLInputElement, TextFieldProps> =
  forwardRef(
    (
      {
        color = 'primary',
        variant = 'filled',
        label,
        disabled = false,
        sx,
        ...props
      },
      ref,
    ) => {
      const style: ThemeUIStyleObject = {
        ...(!disabled
          ? {
            cursor: 'pointer',
            ...(variant === 'filled' && {
              color: 'surface.onVariant',
              bg: 'surface.main',
              '&:focus-within': {
                cursor: 'text',
                caretColor: `${color}.main`,
                borderWidth: 2,
                borderColor: `${color}.main`,
              },
              '& > input:not(:placeholder-shown) ~ span, & > input:focus ~ span':
                  {
                    order: 1,
                    position: 'initial',
                    pt: 2,
                    justifySelf: 'flex-start',
                    alignSelf: 'flex-start',
                    variant: 'body.sm',
                    color: `${color}.main`,
                  },
              '& > input:not(:placeholder-shown), & > input:focus': {
                pb: 2,
              },
            }),
            // TODO: Write styles for outlined variant
            ...(variant === 'outlined' && {}),
          }
          : {
            opacity: 'disabled',
            color: 'surface.onMain',
            bg: 'transparent',
            userSelect: 'none',
            '&::before': {
              content: '""',
              position: 'absolute',
              top: 0,
              left: 0,
              width: '100%',
              height: '100%',
              opacity: 0.04,
              bg: variant === 'filled' && 'surface.onMain',
            },
            '&:hover::after': {
              content: 'none',
            },
          }),
      };

      return (
        <Box
          as="label"
          sx={{ ...style, ...sx }}
          {...__internalProps({ __css: __labelCss })}
        >
          <Box
            ref={ref}
            as="input"
            placeholder="&nbsp;"
            {...({ disabled, ...props } as BoxProps)}
            {...__internalProps({
              __css: {
                order: 2,
                variant: 'body.lg',
                color: 'surface.onMain',
                bg: 'inherit',
                border: 'none',
                outline: 'none',
                appearance: 'none',
              },
            })}
          />
          {label && (
            <Typography
              as="span"
              color={disabled ? 'surface.onMain' : 'currentColor'}
              {...__internalProps({
                __css: {
                  position: 'absolute',
                  transition: 'emphasized',
                },
              })}
            >
              {label}
            </Typography>
          )}
        </Box>
      );
    },
  );
