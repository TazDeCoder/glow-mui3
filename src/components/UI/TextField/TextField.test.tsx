import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { TextField } from './TextField';

describe('<TextField />', () => {
  describe('Snapshots', () => {
    it('renders correctly', () => {
      const { container } = render(<TextField />);
      expect(container).toMatchSnapshot();
    });
  });
  describe('User events', () => {
    it('displays user typed input', async () => {
      const { getByRole } = render(<TextField />);
      const textboxEl = getByRole('textbox');
      await userEvent.click(textboxEl);
      await userEvent.type(textboxEl, 'Hello, world!');
      expect(textboxEl).toHaveValue('Hello, world!');
    });
  });
});
