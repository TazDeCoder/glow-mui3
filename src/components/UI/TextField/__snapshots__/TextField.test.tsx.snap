// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> Snapshots renders correctly 1`] = `
.emotion-0 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: -webkit-min-content;
  width: -moz-min-content;
  width: min-content;
  height: 56px;
  padding-left: 16px;
  padding-right: 16px;
  overflow: hidden;
  background-color: surface.variant;
  border-bottom: 1px solid currentColor;
  border-top-left-radius: xs;
  border-top-right-radius: xs;
  cursor: pointer;
  color: surface.onVariant;
  background-color: surface.main;
}

.emotion-0:hover::after {
  content: "";
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background-color: action.hover;
  border-radius: inherit;
}

.emotion-0:focus-within {
  cursor: text;
  caret-color: primary.main;
  border-width: 2px;
  border-color: primary.main;
}

.emotion-0>input:not(:placeholder-shown)~span,
.emotion-0>input:focus~span {
  -webkit-order: 1;
  -ms-flex-order: 1;
  order: 1;
  position: initial;
  padding-top: 8px;
  justify-self: flex-start;
  -webkit-align-self: flex-start;
  -ms-flex-item-align: flex-start;
  align-self: flex-start;
  color: primary.main;
}

.emotion-0>input:not(:placeholder-shown),
.emotion-0>input:focus {
  padding-bottom: 8px;
}

.emotion-1 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  -webkit-order: 2;
  -ms-flex-order: 2;
  order: 2;
  color: surface.onMain;
  background-color: inherit;
  border: none;
  outline: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
}

<div>
  <label
    class="emotion-0"
  >
    <input
      class="emotion-1"
      placeholder="Â "
    />
  </label>
</div>
`;
